File: ai-service/app/core/vault.py
"""Vault client for reading secrets from HashiCorp Vault."""
import os
import hvac
from functools import lru_cache
from typing import Dict, Any, Optional


class VaultClient:
    """HashiCorp Vault client for secret management."""
    
    def __init__(self):
        self.vault_addr = os.getenv("VAULT_ADDR", "http://localhost:8200")
        self.role_id = os.getenv("AI_SERVICE_ROLE_ID")
        self.secret_id = os.getenv("AI_SERVICE_SECRET_ID")
        self.client: Optional[hvac.Client] = None
        self._authenticated = False
        
    def authenticate(self) -> bool:
        """Authenticate with Vault using AppRole."""
        if not self.role_id or not self.secret_id:
            raise ValueError(
                "AI_SERVICE_ROLE_ID and AI_SERVICE_SECRET_ID must be set"
            )
        
        try:
            self.client = hvac.Client(url=self.vault_addr)
            
            # Authenticate using AppRole
            response = self.client.auth.approle.login(
                role_id=self.role_id,
                secret_id=self.secret_id,
            )
            
            self._authenticated = True
            return True
            
        except Exception as e:
            raise ConnectionError(f"Failed to authenticate with Vault: {e}")
    
    def read_secret(self, path: str) -> Dict[str, Any]:
        """
        Read a secret from Vault KV v2 engine.
        
        Args:
            path: Secret path without mount point (e.g., "ai-service/api-keys")
            
        Returns:
            Dictionary containing secret data
        """
        if not self._authenticated:
            self.authenticate()
        
        try:
            # Read from KV v2 (offgrid mount point)
            secret = self.client.secrets.kv.v2.read_secret_version(
                path=path,
                mount_point="offgrid"
            )
            
            return secret["data"]["data"]
            
        except Exception as e:
            raise ValueError(f"Failed to read secret from {path}: {e}")
    
    def get_ai_credentials(self) -> Dict[str, str]:
        """Get AI service API keys from Vault."""
        return self.read_secret("ai-service/api-keys")
    
    def is_healthy(self) -> bool:
        """Check if Vault connection is healthy."""
        try:
            if not self.client:
                self.client = hvac.Client(url=self.vault_addr)
            return self.client.sys.is_initialized() and not self.client.sys.is_sealed()
        except Exception:
            return False


@lru_cache()
def get_vault_client() -> VaultClient:
    """Get singleton Vault client instance."""
    client = VaultClient()
    client.authenticate()
    return client

File: frontend/src/lib/vault.ts
// Vault client for server-side secret management
import vault from 'node-vault';

interface VaultConfig {
  apiVersion: string;
  endpoint: string;
  token?: string;
}

interface VaultSecrets {
  [key: string]: string;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type VaultClientType = any;

class VaultClient {
  private client: VaultClientType;
  private roleId: string;
  private secretId: string;
  private authenticated: boolean = false;

  constructor() {
    const vaultAddr = process.env.VAULT_ADDR || 'http://localhost:8200';
    this.roleId = process.env.FRONTEND_ROLE_ID || '';
    this.secretId = process.env.FRONTEND_SECRET_ID || '';

    const config: VaultConfig = {
      apiVersion: 'v1',
      endpoint: vaultAddr,
    };

    this.client = vault(config);
  }

  async authenticate(): Promise<void> {
    if (!this.roleId || !this.secretId) {
      throw new Error(
        'FRONTEND_ROLE_ID and FRONTEND_SECRET_ID must be set'
      );
    }

    try {
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access
      const response = await this.client.approleLogin({
        role_id: this.roleId,
        secret_id: this.secretId,
      });

      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access
      this.client.token = response.auth.client_token;
      this.authenticated = true;
    } catch (error) {
      throw new Error(
        `Failed to authenticate with Vault: ${error instanceof Error ? error.message : 'Unknown error'}`
      );
    }
  }

  async readSecret(path: string): Promise<VaultSecrets> {
    if (!this.authenticated) {
      await this.authenticate();
    }

    try {
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access
      const response = await this.client.read(`offgrid/data/${path}`);
      // eslint-disable-next-line @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-member-access
      return response.data.data;
    } catch (error) {
      throw new Error(
        `Failed to read secret from ${path}: ${error instanceof Error ? error.message : 'Unknown error'}`
      );
    }
  }

  async getFrontendConfig(): Promise<VaultSecrets> {
    return this.readSecret('frontend/config');
  }

  async getStripeKeys(): Promise<VaultSecrets> {
    return this.readSecret('frontend/stripe');
  }

  async getAICredentials(): Promise<VaultSecrets> {
    return this.readSecret('ai-service/api-keys');
  }

  async isHealthy(): Promise<boolean> {
    try {
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access
      const health = await this.client.health();
      // eslint-disable-next-line @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-member-access
      return health.initialized && !health.sealed;
    } catch {
      return false;
    }
  }
}

// Singleton instance
let vaultClient: VaultClient | null = null;

export function getVaultClient(): VaultClient {
  if (!vaultClient) {
    vaultClient = new VaultClient();
  }
  return vaultClient;
}

export default VaultClient;
